### ISSUE TITLE
Obtain uniprot annotations for S.salar genes

### ISSUE THREAD

**Issue by shellytrigg** [RobertsLab/resources#1220]
@kubu4 can you create an annotations file (like this one https://raw.githubusercontent.com/hputnam/Geoduck_Meth/master/RAnalysis/Data/Genome/Panopea-generosa-genes-annotations.tab) for the genes in this Atlantic salmon gff (https://gannet.fish.washington.edu/metacarcinus/Salmo_Calig/GENOMES/v2/RefSeq/GCF_000233375.1_ICSASG_v2_genomic.gff)?

**Comment by kubu4** [comment]
Ha! This file again...

I know we had lengthy discussions related to this file and the OSF repo, but I can't seem to track any of them down.

@sr320 I think you generated the `Panopea-generosa-genes-annotations.tab`. Would you happen to have a link to a notebook entry I could use for this?

**Comment by kubu4** [comment]
Aaaaaaand, of course after I post that, I immediately found the discussion! It was Hollie's Geoduck_Meth repo (ended up searching Gmail to find the discussion).

https://github.com/hputnam/Geoduck_Meth/issues/16#issuecomment-786848374

I'll try to get something whipped up!

**Comment by sr320** [comment]
@shellytrigg - remind me what end goal with this effort is?

**Comment by shellytrigg** [comment]
@sr320 these annotations will be merged with genes that contain DMRs or are near DMRs. Then these genes can be compared with the genes that came out of the DEG analysis. The goal of this analysis is to be a gene-centric approach that is different but complementary to the approach that Cristian is taking with overlaying the meth data and expression data in the circos plot

**Comment by sr320** [comment]
Gotcha - but with ducks this was based on our blast - I presume with salmon thes ID have already been generated by NCBI etc.... Main thing is we need to confirm we use what Cris uses..

**Comment by shellytrigg** [comment]
Cris confirmed he is using that same gff file linked above. Here is the list of DEGs and the name column contains either the LOC geneID or a gene name that I could map to Alternatively, there are genomic locations listed in this file that I could map to. 

I was also interested in getting the GO terms for these genes which is why I wanting to use Uniprot, so that some functional inferences can be made after

**Comment by shellytrigg** [comment]
To clarify, the GeneID (LOC numbers) and gene names are assigned in the gff file that is linked above and downloaded from NCBI. But not every gene has a gene name and there are no GO IDs listed. So the goal with Uniprot mapping was to try to get any missing gene names and GO annotations

**Comment by sr320** [comment]
@kubu4 where does this stand?

**Comment by kubu4** [comment]
Actively working on it (still). I'm probably trying to "over-engineer" it o make it reproducible, but it's a pain. This is in part because some of the genes do not have any matches in UniProt, while others have multiple matches... It's also related to trying to learn/use Python to convert GFF accessions to UniProt IDs to then use for my script to retrieve GO terms.

I should probably just resort to using the web interface to retrieve GO info and then worry about joining stuff...

**Comment by kubu4** [comment]
Here's the current approach I've been using, in case anyone wants to weigh in:

1. Extract NCBI gene ids (listed in Dbxref attribute) using GFFutils (this is easy).
2. Map NCBI gene ids to UniProt accessions via Python (technically easy since UniProt provides Python script, but becomes tricky when trying to handle no matches and multiple matches; plus learning how to manipulate outputs (e.g. strings vs lists) is a bit challenging for me).
3. Use UniProt accessions to pull GO terms, using an existing bash script (see `uniprot2go.sh` in code repo). This works, but specifically requires UniProt accessions.
4. Join all of this junk into desired tab-delimited file...

**Comment by kubu4** [comment]
Update. I actually have the Python code running successfully! Getting closer...

**Comment by kubu4** [comment]
Very close now... Parsing is working properly. Next is to do some joining. 

A usable file should be available before noon PST today.

**Comment by kubu4** [comment]
Well, script is still running... I'll post link when it's finished. But, I've already done all the testing and things look like they will be correctly formatted/matched up when it does finish...

**Comment by kubu4** [comment]
Well, that took ~6.5hrs to run! Not what I was expecting. Anyway, here's the file:

https://gannet.fish.washington.edu/Atumefaciens/20210601_ssal_gff-annotations/20210601_ssal_chrom-gene_id_start-end-acc-gene-gene_description-go_ids.tab

Jupyter Notebook here:

https://nbviewer.jupyter.org/github/RobertsLab/code/blob/master/notebooks/sam/20210601_ssal_gff-annotations.ipynb

_Please_ look through the Jupyter Notebook and review my process!! I tried hard to make sure I checked things at each step (I know I had to tweak things many times when I was double-checking things), but there's a lot going on here.

Post any questions here!

Also, I did just realize that this includes UniProt "reviewed" and "_unreviewed_". It won't be difficult to tweak the job I ran to only include "reviewed". Just let me know.
